working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,559 @@
+      # Act
+      result = validate_book(book_id)
+
+      # Assert
+      assert result.id == book_id
+
+  def test_validate_book_with_invalid_id():
+      # Arrange
+      book_id = "cat"
+
+      # Act & Assert
+      with pytest.raises(Exception) as e:
+          validate_book(book_id)
+          assert str(e.value) == "book cat invalid"
+
+  def test_validate_book_with_nonexistent_id():
+      # Arrange
+      book_id = 3
+
+      # Act & Assert
+      with pytest.raises(Exception) as e:
+          validate_book(book_id)
+          assert str(e.value) == "book 3 not found"
+  ```
+</details>
+
+#### Testing `update_book`
+
+With `validate_book` we were lucky and Copilot outlined all of the scenarios that we intended to test, but that will not always be the case. We won't go step by step again, but we want to take the same initial approach to testing our remaining dependencies `update_book` & `delete_book`. 
+
+Let's use the Copilot inline chat to ask for help generating tests for the `update_book` function and take a look at what Copilot suggested below:
+```
+def test_update_book(client, two_saved_books):
+    # Arrange
+    book_id = 1
+    data = {
+        "title": "Updated Book",
+        "description": "New description"
+    }
+
+    # Act
+    response = client.put(f"/books/{book_id}", json=data)
+    response_body = response.get_json()
+
+    # Assert
+    assert response.status_code == 200
+    assert response_body == "Book #1 successfully updated"
+
+def test_update_book_invalid_id(client, two_saved_books):
+    # Arrange
+    book_id = "cat"
+    data = {
+        "title": "Updated Book",
+        "description": "New description"
+    }
+
+    # Act
+    response = client.put(f"/books/{book_id}", json=data)
+    response_body = response.get_json()
+
+    # Assert
+    assert response.status_code == 400
+    assert response_body == {"message": "book cat invalid"}
+
+def test_update_book_nonexistent_id(client, two_saved_books):
+    # Arrange
+    book_id = 3
+    data = {
+        "title": "Updated Book",
+        "description": "New description"
+    }
+
+    # Act
+    response = client.put(f"/books/{book_id}", json=data)
+    response_body = response.get_json()
+
+    # Assert
+    assert response.status_code == 404
+    assert response_body == {"message": "book 3 not found"}
+```
+
+This time Copilot did a great job writing tests that properly use our test fixtures and assert on the correct values! The tests above will run as-suggested, and we have a good range of scenarios:
+- successfully update a book
+- return an error if the book id is invalid
+- return an error if a book with the id doesn't exist
+
+However, there are other scenarios that are worth testing. What happens if the request dictionary is missing either the title or description keys? How will our code behave if the user adds extra keys to the dictionary in the request? We can open up the Copilot chat and ask for help generating the tests for these scenarios. 
```

Comment:
- made the keys strings
- strictly speaking could leave "the" but "the Copilot" sounded weird, even though "the" is applying more to "chat"
- trailing spaces
```suggestion
However, there are other scenarios that are worth testing. What happens if the request dictionary is missing either the `"title"` or `"description"` keys? How will our code behave if the user adds extra keys to the dictionary in the request? We can open up Copilot chat and ask for help generating the tests for these scenarios.
```